### PR TITLE
(1138) Paginate case list

### PR DESCRIPTION
**Pull request by yndajas**
## Context

Taking the foundations from #86. Another reference PR could be https://github.com/ministryofjustice/hmpps-approved-premises-ui/pull/1010 but there's a lot going on there

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [ ] There are changes required to the Accredited Programmes API for this change to work...
  - [ ] ... and they have been released to production already

### Post-merge

- [ ] This adds, extends or meaningfully modifies a feature...
  - [ ] ... and I have written or updated an end-to-end test for the happy path in the [Accredited Programmes E2E repo](https://github.com/ministryofjustice/hmpps-accredited-programmes-e2e)
- [ ] This makes new expectations of the API...
  - [ ] ... and I have notified the API developer(s) of changes to the contract tests (Pact), or the API is already compliant
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

File: server/data/accreditedProgrammesApi/referralClient.test.ts [hunk 1]
```diff
@@ -120,20 +121,25 @@ pactWith({ consumer: 'Accredited Programmes UI', provider: 'Accredited Programme
...
-        totalPages: 1,
+        pageNumber: 1,
+        pageSize: 15,
+        totalElements: 16,
+        totalPages: 2,
       }
 
       beforeEach(() => {
         provider.addInteraction({
+          // DOES THIS NOT CONFLICT WITH THE STATE ABOVE? 1 vs 16 with specific params
```
Comment (Gweaton) [inline server/data/accreditedProgrammesApi/referralClient.test.ts:138]: I'm not sure, but we can test this one out locally against the API and see how it behaves

File: server/services/referralService.ts [hunk 2]
```diff
@@ -40,7 +40,7 @@ export default class ReferralService {
   async getReferralSummaries(
     username: Express.User['username'],
     organisationId: Organisation['id'],
-    filterValues?: { audience?: CourseAudience['value']; status?: string },
+    filterValues?: { audience?: CourseAudience['value']; page?: string; status?: string },
```
Comment (Gweaton) [inline server/services/referralService.ts:43]: Part of me is wondering whether `page` should be a separate argument outside of `filterValues` but I suppose we're technically still filtering down data, so I think this is fine.

File: server/utils/referrals/caseListUtils.ts [hunk 3]
```diff
@@ -52,6 +61,57 @@ export default class CaseListUtils {
...
+    path: Request['path'],
+    queryParamsExcludingPage: Array<QueryParam>,
+    pageNumber: Paginated<ReferralSummary>['pageNumber'],
+    totalPages: Paginated<ReferralSummary>['totalPages'],
+  ): GovukFrontendPaginationWithItems {
+    const oneIndexedPageNumber = pageNumber + 1
+
+    const pagination: GovukFrontendPaginationWithItems = { items: [] }
+
+    const numberPageLinkNumbers = Array.from(
```
Comment (Gweaton) [inline server/utils/referrals/caseListUtils.ts:74]: I'm finding this variable name a bit hard to parse. Is this getting the links for each page? I wonder if breaking this down into smaller variables might help improve readability here e.g. storing the `Set` value in a variable. Would `pageLinkNumbers` or even `pageNumbers` be enough here?

File: server/utils/referrals/caseListUtils.ts [hunk 4]
```diff
@@ -52,6 +61,57 @@ export default class CaseListUtils {
...
+      new Set([1, oneIndexedPageNumber - 1, oneIndexedPageNumber, oneIndexedPageNumber + 1, totalPages]),
+    ).filter((number: Paginated<ReferralSummary>['pageNumber']) => number > 0 && number <= totalPages)
+
+    numberPageLinkNumbers.forEach((numberPageLinkNumber, numberPageLinkNumberIndex) => {
+      if (
+        numberPageLinkNumber !== 1 &&
```
Comment (Gweaton) [inline server/utils/referrals/caseListUtils.ts:80]: again, although we'd be declaring variables here and re-using them, I think it'd help readability if we could encapsulate some of this logic in a clear variable name to make this a bit more reader friendly

**Review by Gweaton — COMMENTED** [review]
Looking good on first review - the pagination function is rather complex and I've found it a bit hard to understand. I think being really explicit with names or extracting into private functions where possible would help with this.

**Comment by Gweaton** [comment]
I've just tested this running against dev locally and I've got a few blank pages at the end: 
image

File: server/utils/referrals/caseListUtils.ts (hunk 3 above)
Comment (yndajas) [reply server/utils/referrals/caseListUtils.ts:74]: `pageNumbers`

File: server/utils/referrals/caseListUtils.ts [hunk 5]
```diff
@@ -52,6 +61,57 @@ export default class CaseListUtils {
...
+        item.current = true
+      }
+
+      pagination.items.push(item)
+    })
+
+    if (oneIndexedPageNumber > 1) {
+      pagination.previous = {
+        href: CaseListUtils.paginationLinkUrl(path, queryParamsExcludingPage, oneIndexedPageNumber - 1),
+        text: 'Previous',
```
Comment (yndajas) [inline server/utils/referrals/caseListUtils.ts:101]: This text is not needed: the default has "page" visually hidden, so this removes that

**Comment by yndajas** [comment]
Replaced by #408 and #409
